cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DataprocProvisioner.java

```diff
@@ -152,6 +168,114 @@ public Cluster createCluster(ProvisionerContext context) throws Exception {
     }
   }
 
+  private Map<? extends String, ? extends String> getReuseLabels(ProvisionerContext context, DataprocConf conf) {
+    Map<String, String> reuseLabels = new HashMap<>();
+    String normalProfileName = getNormalizedProfileName(context);
+    if (normalProfileName != null) {
+      reuseLabels.put(LABEL_PROFILE, normalProfileName);
+    }
+    reuseLabels.put(LABEL_REUSE_KEY, conf.getClusterReuseKey());
+    reuseLabels.put(LABEL_REUSE_FOR, getClusterName(context));
+    return reuseLabels;
+  }
+
+  /**
+   * If cluster reuse is enabled & possible tries to find a cluster to reuse.
+   * @param client data proc client
+   * @param context provisioner contex
+   * @param conf dataproc configuration
+   * @return a cluster ready to reuse or null if none available.
+   */
+  @Nullable
+  private Cluster tryReuseCluster(DataprocClient client, ProvisionerContext context, DataprocConf conf)
+    throws RetryableProvisionException, IOException {
+    if (!isReuseSupported(conf)) {
+      LOG.debug("Not checking cluter reuse, enabled: {}, skip delete: {}, idle ttl: {}, reuse threshold: {}",
+                conf.isClusterReuseEnabled(), conf.isSkipDelete(), conf.getIdleTTLMinutes(),
+                conf.getClusterReuseThresholdMinutes());
+      return null;
+    }
+
+    String defaultName = getClusterName(context);
+
+    //For idempotency, check if we already have the cluster allocated
+    Optional<Cluster> clusterOptional = getAllocatedCluster(defaultName, client);
+    if (clusterOptional.isPresent()) {
+      Cluster cluster = clusterOptional.get();
+      if (cluster.getStatus() == ClusterStatus.CREATING ||
+        cluster.getStatus() == ClusterStatus.RUNNING) {
+        LOG.debug("Found allocated cluster {}", cluster.getName());
+        return cluster;
+      } else {
+        LOG.debug("Preallocated cluster {} expired, will look for new one");
+        //Let's remove the reuse label to ensure new cluster will be picked up by getAllocatedCluster
+        try {
+          client.updateClusterLabels(cluster.getName(), Collections.emptyMap(), Collections.singleton(LABEL_REUSE_FOR));
+        } catch (Exception e) {
+          LOG.debug("Error during cluster cleanup, probably cluster died already", e);
+        }
+      }
+    }
+
+    Lock reuseLock = getSystemContext().getLock(REUSE_LOCK);
+    reuseLock.lock();
+    try {
+      Map<String, String> filter = new HashMap<>();
+      String normalizedProfileName = getNormalizedProfileName(context);
+      if (normalizedProfileName != null) {
+        filter.put(LABEL_PROFILE, normalizedProfileName);
+      }
+      filter.put(LABEL_VERSON, getVersionLabel());
```

Comment:
do we need to filter on the version? It's because if somebody upgrades we want the pipelines not to re-use the existing cluster?

Reply:
Yes. It's directly after upgrade. Some jobs may be running. I don't want jobs from new versions to reuse cluster from old versions as there can be provisioner differences.